Ingest code refactoring

zhongh opened the pull request:
The new ingest code structure is comprised of 4 pieces:
1) A very short ingest-xxxxx-refactored.py for the ingest process, users need to make a copy of it and change certain variable values (file paths, index and type names, etc) in it to fit certain cases
2) an ingestHelpersRefactored.py storing some independent helper functions; users will need to create their own create_xxxxx_doc function
3) Ingest.py, a class governing the ingest process
4) Maybe.py, auxilary class for those helper functions in ingestHelpersRefactored.py

How does it look like? 

Next, if this works, we can move on to generalize it. For example, say, for datasets we will need the author and the author's attribute A, B, and C, while for publications we also need the author, but authors attribute A, C, and D. Then a get_author function getting attributes A, B, C, and D can be shared by both ingest processes, since what will eventually be in the output will be limited by describeXXXX.rq and the mapping file. Does it make sense?

zhongh committed:
Debugged, refactored ingest scripts; the main script to run is ingest-datasets-refactored.py, with the helper script called ingestHelpersRefactored.py and class file Ingest.py

zhongh committed:
Minor restructures, fixes and comments added; ready for review on overall code structure to see if we can move on in generalize it.

mrpatrickwest commented:
You've created a branch for the code. The names of the new .py files don't need to have refactored at the end. Just call them what they are.

ingest-dataset.py and ingestHelpers.py

mrpatrickwest commented:
Question ... why is the class called Maybe? Is it a utility class? What's it for? There are no comments on the class or the methods. I want comments!

mrpatrickwest commented on ingest/ingest-datasets-refactored.py:
```diff
@@ -0,0 +1,72 @@
+
+# Finally, run the ingest process. See explanations and examples below.
+
+# Usage:
```
Please remove the commented out code. We don't need it any more.

zhongh replied:
Which part, those comments about usage?

mrpatrickwest replied:
Oh ... I see. These comments will be replicated then between the different ingest scripts. Wondering if we should add these to a README file in this directory instead of in code. Or add it as part of the main README.md under an ingest section? @zednis what do you think?

zhongh replied:
I suggest adding a README in this directory? Since it is just comments on how to create and run customized ingest process.

mrpatrickwest commented on ingest/ingestHelpersRefactored.py:
```diff
@@ -0,0 +1,292 @@
+         create_object_doc_function, object_index, object_type, variable_name_sparql):
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--threads', default=2, help='number of threads to use (default = 8)')
```
change the default to 4, and change the help string for the default.

zednis commented:
@mrpatrickwest I'll add the comments to the class Maybe as I wrote it.  The class is an implementation of  the optional datatype - a container that may or may not contain a value.  It additionally provides methods for common functional programming functions (map, flatMap, filter, etc) on the object (if any) contained in the Maybe object.  

Why do this?  So we can use functional programming concepts in python with a fluent interface.

This class was heavily influenced by functional programming concepts and the Java 8 Optional and Stream interfaces.

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html
https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html

I'll add the comments, but I would rather do so in a separate branch.  I can make this a task for next sprint if you wish.

mrpatrickwest commented:
@zednis ... works for me.

zhongh commented:
1) Some addition comments on Maybe.py: In getting those more complicatedly structured values, for instance the "author" and "distribution", I didn't utilize the class Maybe (see get_authors and get_distributions). Instead I just used the primitive way to do it, which was pretty intuitive for me during development.

2) Does the whole framework look okay to move on? I prefer deleting the -refactored suffixes and substitute the original file when the new framework seems good to you...... Just in case we need to move backward somewhere.

zednis commented:
As @mrpatrickwest said, we do not need to add 'refactored' to the class names.

zhongh committed:
20151022: Remove the -refactored suffixes; changed default number of threads.

Commit:
Trivial edits.

mrpatrickwest commented on ingest/ingest-datasets.py:
```diff
@@ -1,148 +1,72 @@
+
+# Finally, run the ingest process. See explanations and examples below.
+
+# Usage:
```
From line 32 to 72 these are general comments about the ingest process. General usage, sample command line for executing commands against elasticsearch.

Do you think it would be a good idea to create a README with this?

zhongh replied:
Yes I think it's better that these should be in a README

On Friday, October 23, 2015, Patrick West notifications@github.com wrote:

> In ingest/ingest-datasets.py
> https://github.com/tetherless-world/dco-elasticsearch/pull/29#discussion_r42831500
> :
> 
> > \ No newline at end of file
> > -    Main(get_objects_query_location=GET_OBJECTS_QUERY_LOCATION,
> > -         describe_object_query_location=DESCRIBE_DATASET_OBJECT_LOCATION,
> > -         create_object_doc_function=create_dataset_doc,  # Need to create case-varying create_xxxxxx_doc in helper file
> > -         object_index=INDEX, object_type=TYPE,
> > -         variable_name_sparql=VARIABLE_NAME_SPARQL)
> >   +
> >   +# Finally, run the ingest process. See explanations and examples below.
> >   +
> >   +# Usage:
> 
> From line 32 to 72 these are general comments about the ingest process.
> General usage, sample command line for executing commands against
> elasticsearch.
> 
> Do you think it would be a good idea to create a README with this?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/tetherless-world/dco-elasticsearch/pull/29/files#r42831500
> .

zednis replied:
:+1:

zhonghao1014 replied:
Should the file be a README.md or just just README

On Friday, October 23, 2015, Stephan Zednik <notifications@github.com>
wrote:

> In ingest/ingest-datasets.py
> <https://github.com/tetherless-world/dco-elasticsearch/pull/29#discussion_r42832564>
> :
>
> > \ No newline at end of file
> > +    Main(get_objects_query_location=GET_OBJECTS_QUERY_LOCATION,
> > +         describe_object_query_location=DESCRIBE_DATASET_OBJECT_LOCATION,
> > +         create_object_doc_function=create_dataset_doc,  # Need to create case-varying create_xxxxxx_doc in helper file
> > +         object_index=INDEX, object_type=TYPE,
> > +         variable_name_sparql=VARIABLE_NAME_SPARQL)
> > +
> > +# Finally, run the ingest process. See explanations and examples below.
> > +
> > +# Usage:
>
> [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tetherless-world/dco-elasticsearch/pull/29/files#r42832564>
> .
>

zhongh replied:
Should the file name be README.md or just README? Thanks!

zednis replied:
README.md and use markdown syntax

https://help.github.com/articles/github-flavored-markdown/

zhongh replied:
Sure. Just pushed a README now will change it with markdown syntax.

zhongh committed:
Comments in ingest-datasets.py updated; usage and examples of ingest process moved to ingest/README.md; further adjustments in the .md file might be needed.

zhongh committed:
Merge branch 'ingest-code-refactoring' of https://github.com/tetherless-world/dco-elasticsearch into ingest-code-refactoring

zhongh committed:
Some formatting on READMD.md

zhongh committed:
Further changes in markdowns in README

zhongh committed:
Further again changes in README.md

zhongh committed:
Further again changes in README.md

zhongh committed:
Further almost-there changes in README.md

zhongh committed:
Nearly-final changes in README.md

Commit:
README.md done

Sorry for the huge number of commits... I could have just done it in the browser and "preview" it while editing!

mrpatrickwest commented on ingest/Ingest.py:
```diff
@@ -0,0 +1,102 @@
+import requests
+
+class Ingest:
+    """say something"""
```
What is this?

zednis replied:
This is a class docstring.  The value isn't helpful, but the string immediately following a class or function defintions is a python docstring.

http://epydoc.sourceforge.net/docstrings.html

mrpatrickwest commented on ingest/Ingest.py:
```diff
@@ -0,0 +1,102 @@
+        self.variable_name_sparql = variable_name_sparql
+
+
+    def load_file(filepath):
```
How difficult is it to add comments to the methods? Use pydoc.

zhongh replied:
So I am now trying to add some neccesary comments.. Could I just use the normal way of adding some comments, instead of using docstring or pydoc?

mrpatrickwest replied:
I would like for you to use pydoc.

zhongh replied:
Check if the newest push looks okay. In this way, we can always use pydoc.help to generate the full documentation of it.

(I am not sure if this is the way you want though.. But at least now they are all written down so it should be easy to change ways to present..)

zhongh replied:
Running pydoc.help(Ingest.py) will generate:

![image](https://cloud.githubusercontent.com/assets/11859381/10871619/fa9a5b98-80b9-11e5-9045-814f40908fde.png)

zednis replied:
Let's stop using the term `object` in the variable names and documentation as the term is very ambiguous.

zhongh replied:
Let's start from this term: Indeed, I totally agree, using the term
"object" as a generalized substitution for different cases such as
dataset/publication/project/etc seems a bit ambiguous. What is a good term
that you suggest? "x"? "thing"? I suggest that we nail the name of this
term down first so that we don't have to change again.

On Sunday, November 1, 2015, Stephan Zednik notifications@github.com
wrote:

> In ingest/Ingest.py
> https://github.com/tetherless-world/dco-elasticsearch/pull/29#discussion_r43588644
> :
> 
> > +import requests
> > +
> > +class Ingest:
> > -    """say something"""
> >   +
> > -    def **init**(self, object_index, object_type,
> > -                 get_objects_query_location, describe_object_query_location, variable_name_sparql):
> > -        self.object_index = object_index
> > -        self.object_type = object_type
> > -        self.records = []
> > -        self.get_objects_query = Ingest.load_file(get_objects_query_location)
> > -        self.describe_object_query = Ingest.load_file(describe_object_query_location)
> > -        self.variable_name_sparql = variable_name_sparql
> >   +
> >   +
> > -    def load_file(filepath):
> 
> Let's stop using the term object in the variable names and documentation
> as the term is very ambiguous.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/tetherless-world/dco-elasticsearch/pull/29/files#r43588644
> .

zhongh committed:
Documentation in Ingest.py updated; please check if this way works.

zednis commented on ingest/Ingest.py:
```diff
@@ -0,0 +1,162 @@
+                 get_objects_query_location, describe_object_query_location, variable_name_sparql):
+        """
+        Constructor method of the Ingest class.
+        :param object_index:                    object index
```
changes this to something more descriptive such as `elasticsearch_index`

edit: also provide a more descriptive _description_ in the documentation than a repeat of the variable name.  I had to look at the usage to understand what `object_index` was because that is a very ambiguous name.

zhonghao1014 replied:
while we change object_index -> elasticsearch_index, should we also change:
object_type -> elasticsearch_type?

zednis commented on ingest/Ingest.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        Constructor method of the Ingest class.
+        :param object_index:                    object index
+        :param object_type:                     object type
```
change this to something more descriptive such as `elasticsearch_type`

zednis commented on ingest/ingestHelpers.py:
```diff
@@ -141,119 +151,142 @@ def get_distributions(ds):
+#    Implementation of Main(...):
+#
 
+def Main(get_objects_query_location, describe_object_query_location,
```
don't start methods with capital letters - that is reserved in python for class definitions.

zednis commented on ingest/ingest-datasets.py:
```diff
@@ -1,148 +1,34 @@
-
-    Main(get_objects_query=get_datasets_query, create_object_doc_function=create_dataset_doc,
-         object_index=_index, object_type=_type)
+    Main(get_objects_query_location=GET_OBJECTS_QUERY_LOCATION,
```
Why is the dataset-specific code not in ingest-datasets.py?

zednis commented on ingest/ingestHelpers.py:
```diff
@@ -1,20 +1,15 @@
 __author__ = 'Hao'
```
What is this file?  This has a dataset specific-usage of Ingest yet is called ingestHelpers?  You should probably have the ingest-datasets script contain a class DatasetIngest which extends Ingest and includes the dataset-specific logic.

zhongh replied:
Thank you so much for this advice! That will definitely make it way more organized than now!

zhongh committed:
20151101: Ingest code further refactored; All case-related usage are in ingest-datasets.py, with definition of DatasetIngest, inheriting the base class Ingest, with class method create_x_doc overwritten as appropriate to use case; ingestHelpers.y only contains independent helper functions; other minor changes in variable names are made.

zhongh committed:
20151101-2: Minor recorrections regarding the previous commit.

zhongh commented:
Please check the modified constructions in the lastest commit.

mrpatrickwest commented:
@zhonghao1014 you tried the ingest out on your local machine and it works?

I'll have to defer to @zednis about the new changes you've made, though they look okay to me. I like the fact that we're creating a subclass of Ingest to do the work. Very nice.

zednis commented on ingest/ingestHelpers.py:
```diff
@@ -139,121 +125,44 @@ def get_distributions(ds):
-    # parser.add_argument('--es', default="https://dcotest.tw.rpi.edu/search", help="elasticsearch service URL")
     parser.add_argument('--publish', default=False, action="store_true", help="publish to elasticsearch?")
     parser.add_argument('--rebuild', default=False, action="store_true", help="rebuild elasticsearch index?")
     parser.add_argument('--mapping', default="mappings/dataset.json", help="dataset elasticsearch mapping document")
```
The default value for --mapping should come from the Ingest subclass.

zhongh replied:
@mrpatrickwest @zednis Yes, all pushed commits are tested for both generating bulk file and publishing results.

I will go ahead to make changes to make the default mapping value subclass-based.

zhonghao1014 replied:
I found the easier way to do this is to have mapping passed in as an argument in the constructor. So when we create the instance of a subclass, we can simply pass in the value of the mapping file location.

Then, when creating a specific ingest code, one can just directly change the mapping location among:
GET_OBJECTS_QUERY_LOCATION = "queries/listDatasets.rq"
DESCRIBE_DATASET_OBJECT_LOCATION = "queries/describeDataset.rq"
MAPPING_FILE_LOCATION = "mappings/dataset.json"
INDEX = "dco"
TYPE = "dataset"
VARIABLE_NAME_SPARQL = "?dataset"

What do you think?

zhonghao1014 replied:
Well, we can use the class attributes. But is it having the mapping passed in a more "maneuverable" approach?

zednis commented:
One small suggestion on the default mapping value and I think we will be ready to merge (pending confirmation of successful testing).

zhongh commented:
When I tried to use the --publish, it shows this error message:

```
...
  does not look like a valid URI, trying to serialize this will break.
  does not look like a valid URI, trying to serialize this will break.
Traceback (most recent call last):
  File "ingest-datasets.py", line 114, in <module>
    variable_name_sparql=VARIABLE_NAME_SPARQL, XIngest=DatasetIngest)
  File "/Users/Hao/Projects/dco-elasticsearch/ingest/ingestHelpers.py", line 168, in main
    ingestSomething.publish(bulk=bulk_str, endpoint=args.es, rebuild=args.rebuild, mapping=args.mapping)
  File "/Users/Hao/Projects/dco-elasticsearch/ingest/Ingest.py", line 159, in publish
    r = requests.put(mapping_url, data=mapping_file)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/requests/api.py", line 122, in put
    return request('put', url, data=data, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/requests/sessions.py", line 451, in request
    prep = self.prepare_request(req)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/requests/sessions.py", line 382, in prepare_request
    hooks=merge_hooks(request.hooks, self.hooks),
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/requests/models.py", line 304, in prepare
    self.prepare_url(url, params)
  File "/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/requests/models.py", line 362, in prepare_url
    to_native_string(url, 'utf8')))
requests.exceptions.MissingSchema: Invalid URL 'dcotest.tw.rpi.edu/search//dco/dataset/_mapping': No schema supplied. Perhaps you meant http://dcotest.tw.rpi.edu/search//dco/dataset/_mapping?
```

zednis commented:
@zhonghao1014 have you made a change to your local copy that has not been pushed to the branch yet? i.e. - is the code committed to the branch causing that error?

zhongh commented:
@zednis It's the code already commited to this branch causing this error. But I am not sure if the main branch has this error too. I will try it out..

@zednis (update) I just checked the --publish feature for other ingest code, too. Show the same error message. I guess it must be something wrong with the command I entered...

What is the command you give to let these ingest codes run? Must have been some grammatic mistakes in my commands.

I have been doing the ingest process by manually uploading the mappings and bulk data, that shall be the reason I didn't notice this earlier.

zednis commented:
It looks like requests is throwing an error because it thinks the URL you are trying to connect to is invalid.

error:
`requests.exceptions.MissingSchema: Invalid URL 'dcotest.tw.rpi.edu/search//dco/dataset/_mapping': No schema supplied. Perhaps you meant http://dcotest.tw.rpi.edu/search//dco/dataset/_mapping?`

This looks like the URL passed to requests does not have a protocol?  (e.g. is missing the "http://")

I think you should print out the URL that is being passed to requests.

zhongh committed:
Default value of mapping file location is now specifically defined in subclass DatasetIngest(Ingest).

mrpatrickwest commented on ingest/ingestHelpers.py:
```diff
@@ -139,121 +125,46 @@ def get_distributions(ds):
-        publish(bulk=bulk_str, endpoint=args.es, rebuild=args.rebuild, mapping=args.mapping, index=object_index, tYPE=object_type)
-    print(get_objects_query)
\ No newline at end of file
+        bulk_str = '\n'.join(ingestSomething.records)
```
In a recent branch we added a newline to bulk_str to make sure all documents are added to elasticsearch.

bulk_str = '\n'.join(records)+'\n'

Please make that change

zhongh committed:
Adding the change in other ingest scripts from a recent branch (#34 Added newline to end of bulk file creation for all scripts)

Commit:
Further cleaning up of codes and especially comments; the ambiguous term object replaced by entity representing the entity to be worked on, e.g. dataset, person and etc

zhongh commented:
All features tested and running well.

mrpatrickwest commented:
:shipit:

@zednis ready to ship?

zhongh committed:
Minor changes and documentation fixes for it to work on dcotest; ready to ship.

zhongh commented:
All usage checked on dcotest.tw.rpi.edu. Ready to ship from my side.

mrpatrickwest commented:
looks like you need to merge master into your branch before we can merge. There are conflicts that will need to be resolved. This is what I ended up having to do for one of my projects. IN the top level directory:

git checkout master
git pull
git checkout ingest-code-refactoring
git pull
git merge master

Perhaps some of those steps aren't necessary, but it's what I did and it seemed to work. @zednis ?

resolve the conflicts, re-run your tests to make sure everything still works properly, both on your laptop and then dco-test. Then we can shipit.

zhongh committed:
merge changes from master

zhongh commented:
I first:
git checkout master
git pull
git checkout ingest-code-refactoring
git pull
Then I manually went into the several conflicting files and solved the conflict, then push the commit.
Have tested them.
Ready to ship now.

mrpatrickwest commented:
just checked out this branch and ran it locally. Looks to be working.

:shipit:

zednis commented:
:shipit:
